src/wayland/xwayland_keyboard_grab.rs

```diff
@@ -0,0 +1,197 @@
+use wayland_protocols::xwayland::keyboard_grab::zv1::server::{
+    zwp_xwayland_keyboard_grab_manager_v1::{self, ZwpXwaylandKeyboardGrabManagerV1},
+    zwp_xwayland_keyboard_grab_v1::{self, ZwpXwaylandKeyboardGrabV1},
+};
+use wayland_server::{
+    backend::GlobalId,
+    protocol::{wl_seat, wl_surface},
+    Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+
+use crate::{
+    backend::input::KeyState,
+    input::{
+        keyboard::{self, KeyboardGrab, KeyboardInnerHandle},
+        SeatHandler,
+    },
+    utils::Serial,
+    xwayland::XWaylandClientData,
+};
+
+const MANAGER_VERSION: u32 = 1;
+
+pub trait XWaylandKeyboardGrabHandler: SeatHandler {
+    /// XWayland client has requested a keyboard grab for `surface` on `seat`
+    fn grab(
+        &mut self,
+        surface: wl_surface::WlSurface,
+        seat: wl_seat::WlSeat,
+        grab: XWaylandKeyboardGrab<Self>,
+    );
+
+    fn keyboard_focus_for_xsurface(&self, surface: &wl_surface::WlSurface) -> Self::KeyboardFocus;
```

Comment:
We could have it return an `Option` then log an error if it's `None`. Or if it fails because XWayland is misbehaving, a protocol error would make more sense than crashing the compositor. Though the protocol doesn't define any protocol errors. It doesn't technically even mention that the `wl_surface` should be associated with an X window and mapped. But if it's only used by XWayland, we only need to deal with how XWayland uses the protocol anyway.

Theoretically some sort of race may be possible without `xwayland_shell_v1`

Reply:
Right so `None` and ignoring the grab (which is valid behaviour for a compositor according to the protocol anyway) seems like a good middle-ground. This allows compositor to replace `unwrap` with `?` at least easily, should the surface not be mappable to an `X11Surface`.